clients/richclient/src/main/java/org/openecard/richclient/JnaRuntimeDirectoryFix.java

```diff
@@ -0,0 +1,109 @@
+ * GNU General Public License version 3.0 requirements will be met:
+ * http://www.gnu.org/copyleft/gpl.html.
+ *
+ * Other Usage
+ * Alternatively, this file may be used in accordance with the terms
+ * and conditions contained in a signed written agreement between
+ * you and ecsec GmbH.
+ *
+ ***************************************************************************/
+
+package org.openecard.richclient;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Properties;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class represents a fix that sets the JNA runtime directory to an executable directory, because the default value
+ * /tmp may be mounted as 'noexec' on some systems, which will prevent the startup of the app.
+ *
+ * @author Sebastian Schuberth
+ */
+public class JnaRuntimeDirectoryFix {
+
+    private static final Logger LOG = LoggerFactory.getLogger(JnaRuntimeDirectoryFix.class);
+    private static final String NOEXEC_FLAG = "noexec";
+    private static final String JNA_TMP_DIR = "jna.tmpdir";
+
+    /**
+     * Sets the runtime directory for JNA to an executable directory
+     * <p>
+     * The directory will be set to one of the following, in descending order of priority:
+     * <ul>
+     * <li>the directory pointed to by the System property <em>jna.tmpdir</em>, if this value is set</li>
+     * <li>the <em>/tmp</em> directory, if it is not mounted as 'noexec'. This is also the default for JNA.</li>
+     * <li>the directory pointed to by the enviroment variable <em>XDG_RUNTIME_DIR</em>, if this value is set and the
+     * referenced directory is not mounted as 'noexec'.</li>
+     * <li><em>~/.openecard/run</em> as a fallback if none of the above work
+     * </ul>
+     *
+     * @throws IOException if there is an error reading the "/proc/mounts" file, which contains information about
+     * whether a path is mounted as 'noexec' or not
+     */
+    public static void setJnaRuntimeDirectory() throws IOException {
+	// read value of jna.tmpdir property
+	Properties properties = new Properties(System.getProperties());
+	String propJnaTmpDir = properties.getProperty(JNA_TMP_DIR);
+
+	// if the property has been set externally don't change it
+	if (propJnaTmpDir != null) {
+	    return;
+	}
+
+	//check if we are on Linux
+	String osName = properties.getProperty("os.name");
+	if (osName.contains("nix") || osName.contains("nux") || osName.contains("aix")) {
+
+	    // we are on linux, first read "XDG_RUNTIME_DIR" to see if a user run dir is set
+	    String userRuntimeDir = System.getenv("XDG_RUNTIME_DIR");
+	    boolean canUseNormalTempDir = true;
+	    // if it is not set, we cannot use it
+	    boolean canUseUserRunDir = (userRuntimeDir != null);
+
+	    // then parse "/proc/mounts"
+	    List<MountInfo> mountInfos = MountInfo.getMounts();
+	    for (MountInfo mountInfo : mountInfos) {
+		String mountPath = mountInfo.getMountPath();
+
+		if (mountPath.equals("/tmp")) {
```

Comment:
That should be fixed now with the newest commit. I also found that the construction of the "~/.openecard/run" folder did not work the way I wrote it, as Java does not understand "~", I fixed that as well.

Reply:
I'm not satisfied yet as the new code is not structured according to well defined functionality. Here is a bit of pseudocode how I would say that it should look like.

```
var mounts = readMounts();
if (mounts.hasNoExec(path)) {
  // try other path until a good one is found
}
```

The path stuff should be in the Mounts class
```
hasNoExec(path) {
  // looks up the mount point this path is in
  // will return / if no special path is present or an error if no / is in the list of mountpoints
  var mountPoint = getMatchingMount(path.canonicalize());
  return mountPoint.getFlags().contains("noexec");
}
```

Reply:
I tried to structure the code better now, though I'm not entirely happy with it myself, as I had to resort to a Singleton Pattern. I'm sure this can still be improved.

Reply:
I think this looks much better and is ok this way.